pkg/controller/control.go

```diff
@@ -1244,29 +1248,144 @@ func (c *Controller) handleErrorNoLock(err error) error {
+	// Therefore, it will retain a list of max number of replicas that have the same write bytes RW and other replicas will be set to ERR.
+	// When will len(rwReplicasOnEnospcMap) will not be equal to rwReplicaCount,
+	// For example, there are 3 RW replicas (A,B,C), and 2 RW replicas (B,C) encounter the ENOSPC error.
+	// In this case, rwReplicaCount is 3, and len(rwReplicasOnEnospcMap) is 2, so replicas B and C will be set to ERR, while A will remain RW.
+	if len(rwReplicasOnEnospcMap) == rwReplicaCount {
+		// rwReplicasOnEnospcMap now is a map only contains RW replicas that encountered ENOSPC error.
+		// rwReplicaCount means the count of all RW replicas and if it is not 1 and equal to len(rwReplicasOnEnospcMap),
+		// it means some replica in rwReplicasOnEnospcMap need to be set to ERR for data integrity.
+		if rwReplicaCount != 1 {
+			rwReplicasToErrOnEnospcList := filterReplicasToErrOnEnospc(rwReplicasOnEnospcMap)
+			for _, address := range rwReplicasToErrOnEnospcList {
+				logrus.WithField("volume", c.VolumeName).Infof("Setting RW replica %s to ERR due to no space", address)
+				c.setReplicaModeNoLock(address, types.ERR)
+			}
+		}
+		return true
 	}
-	for address := range replicaNoSpaceErrMap {
+
+	for address := range rwReplicasOnEnospcMap {
 		c.setReplicaModeNoLock(address, types.ERR)
 	}
-	return nil
+
+	return false
+}
+
+// categorizeOutOfSpaceReplicas returns the count of all RW replicas, a map of RW replicas on no space disks, and a list of WO replicas on no space disks.
+// The functions separate RW/WO replicas in replicaWrittenBytesMap into a RW replicas with WriteBytes Map and a WO replicas list to be set to `ERR`.
+//
+//	replicaWrittenBytesMap: a map of replicas that encountered no space error and the bytes they have written when encountering the error. For example: {"replica A": 1, "replica B": 2, "replica C": 2, "replica D": 2, "replica E": 0,}
+func (c *Controller) categorizeOutOfSpaceReplicas(replicaWrittenBytesMap map[string]int) (int, map[string]int, []string) {
+	rwReplicaCount := 0
+	noSpaceRWReplicasMap := make(map[string]int, len(c.replicas))
+	noSpaceWOReplicasList := make([]string, 0, len(c.replicas))
+
+	for _, r := range c.replicas {
+		switch r.Mode {
+		case types.RW:
+			rwReplicaCount++
+			// The RW replicas that are not on out of space disks will not in the replicaWrittenBytesMap.
+			if wbn, exist := replicaWrittenBytesMap[r.Address]; exist {
+				noSpaceRWReplicasMap[r.Address] = wbn
+			}
+		case types.WO:
+			if _, exist := replicaWrittenBytesMap[r.Address]; exist {
+				noSpaceWOReplicasList = append(noSpaceWOReplicasList, r.Address)
+			}
+		}
+	}
+
+	return rwReplicaCount, noSpaceRWReplicasMap, noSpaceWOReplicasList
+}
+
+// filterReplicasToErrOnEnospc filters the replicas that need to be set to ERR on ENOSPC error,
+// and retain max number of replicas that have the same write bytes to keep the volume working.
+//
+//	Returns a list of replicas that need to be set to ERR on ENOSPC error.
```

Comment:
Add the filter rules in the function doc.

Reply:
Added an example to explain the filter rule. Thanks.